### DISCUSSION TITLE
How to be notified when webui is connected

### DISCUSSION THREAD

**Discussion by adauriac** [webui-dev/webui#294] [Q&A] [answered]
Calling webui before its websocket is connected result in an error "Error: WebSocket is not connected"
I would like to call a backend function through webui as soon as possible, is there an event I could listen to to be notified when webui's Websocket is connected?

**Comment by AlbertShown** [comment, marked as answer]
In the backend, using C/C++ for example, you can bind a blank empty ID to get all events, example:

```c
// All events
void events(webui_event_t* e) {
	if (e->event_type == WEBUI_EVENT_CONNECTED) // <- - - - -
		printf("Connected. \n");
	else if (e->event_type == WEBUI_EVENT_DISCONNECTED)
		printf("Disconnected. \n");
	else if (e->event_type == WEBUI_EVENT_MOUSE_CLICK)
		printf("Click. \n");
	else if (e->event_type == WEBUI_EVENT_NAVIGATION) {
		const char* url = webui_get_string(e);
		printf("Starting navigation to: %s \n", url);
		webui_navigate(e->window, url);
	}
}

// Bind all events by using an empty ID
webui_bind(MyWindow, "", events);
```

In the frontend, I guess this option does not exist yet!
But according to [this line](https://github.com/webui-dev/webui/blob/b3a2677809c622af33b687f684d4058c64d8df04/bridge/webui_bridge.ts#L36), please try in JavaScript `webui.wsStatus`

**Reply by adauriac** [reply]
Your solution works, thank you!
But I don't know how to use "webui.wsStatus" in th the frontend to wait for the connection of the websocket, which would be easier.

**Reply by AlbertShown** [reply]
Try (not sure if it will work):
```js
if (webui.wsStatus) {
  // Connected
}
else {
  // Not connected
}
```

However, an easier method should be added in the next coming version.

**Reply by adauriac** [reply]
as far as I can figure out `webui.wsStatus` is always undefined

**Reply by hassandraga** [reply]
Thank you for reporting this. Actually, this is a nice feature to have.

**Reply by adauriac** [reply]
`console.log(webui.#wsStatus)` gives `reference to undeclared private field or method #wsStatus`

**Reply by hassandraga** [reply]
We both noticed that in the same exact moment!
However, I'm going to add this feature today.
Thanks again for the idea.

**Reply by adauriac** [reply]
Thank you for maintaining webui  !

**Reply by hassandraga** [reply]
Clone the latest commit, build the lib. Then add this to your UI:

```js
document.addEventListener('DOMContentLoaded', function() {

	// DOM is loaded, and `webui` object should be available.
	webui.setEventCallback((e) => {

		if (e == webui.event.CONNECTED) {
			// Connection to the backend is established
			console.log('Connected.');
		} else if (e == webui.event.DISCONNECTED) {
			// Connection to the backend is lost
			console.log('Disconnected.');
		}
	});
});
```

**Comment by adauriac** [comment]
Thank you !
I have cloned the github repo.
I 'am (exceptionally) working on windows. In a msys2 windows I launched 
the command mingw32-make.exe
and the directory dist has been created with the dll and lib files.
I also included the piece of code you sent me in my js ui.
My backend is written in python.

At this point I don't  know how to test the new version, keeping in 
place the old one.
Thank you for any advice !

**Reply by hassandraga** [reply]
`webui.setEventCallback` is committed recently to the current main branch. We can't update the Python repo until we are we fully test the current main branch. However, you can manually update your Python package by following those instructions:

1. Download the nightly build from : https://github.com/webui-dev/webui/releases/tag/nightly (_Probably you need `webui-windows-msvc-x64.zip`_)
2. Extract the archive, and copy `webui-2.dll` to your Python package installation folder. Look for `site-packages\webui\webui-windows-msvc-x64` in `AppData\Local\Programs\Python\PythonXXXX\Lib\site-packages` or `C:\msys64\mingw64\lib\python3.X\site-packages`.

After that, you can use `webui.setEventCallback` in your UI.

**Reply by adauriac** [reply]
It works perfectly, thank you !
